locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/softlayer/SoftLayerSameVlanLocationCustomizer.java

```diff
@@ -0,0 +1,385 @@
+        // Check we are customising a SoftLayer location
+        String provider = location.getProvider();
+        if (!(provider.equals(SoftLayerConstants.SOFTLAYER_PROVIDER_NAME))) {
+            String message = String.format("Invalid location provider: %s", provider);
+            LOG.warn(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        // Lookup the latch for this scope
+        String scopeUid = getScopeUid(location);
+        CountDownLatch latch = null;
+        synchronized (lock) {
+            latch = lookupCountDownLatch(location, scopeUid);
+            if (latch == null) {
+                throw new IllegalStateException("No latch available for scope: " + scopeUid);
+            }
+        }
+
+        try {
+            // Check if VLAN IDs have been set for this scope
+            LOG.debug("Looking up saved VLAN details {}", scopeUid);
+            Integer publicVlanId = lookupPublicVlanId(location, scopeUid);
+            Integer privateVlanId = lookupPrivateVlanId(location, scopeUid);
+            if (privateVlanId != null && publicVlanId != null) {
+                LOG.debug("SoftLayer VLANs private {} and public {} already configured for scope: {}",
+                        new Object[] { privateVlanId, publicVlanId, scopeUid });
+                saveVlanDetails(machine, scopeUid, privateVlanId, publicVlanId);
+                return;
+            }
+
+            // No VLAN info yet known, we are probably the first VM and we should
+            // set the VLAN info for others to then learn about.
+
+            // Ask SoftLayer API for the VLAN details for this VM
+            LOG.debug("Requesting VLAN details from API for scope: {}", scopeUid);
+            VirtualGuestApi api = computeService.getContext().unwrapApi(SoftLayerApi.class).getVirtualGuestApi();
+            Long serverId = Long.parseLong(machine.getJcloudsId());
+            VirtualGuest guest = api.getVirtualGuestFiltered(serverId,
+                    "primaryNetworkComponent;" +
+                    "primaryNetworkComponent.networkVlan;" +
+                    "primaryBackendNetworkComponent;" +
+                    "primaryBackendNetworkComponent.networkVlan");
+            publicVlanId = guest.getPrimaryNetworkComponent().getNetworkVlan().getId();
+            privateVlanId = guest.getPrimaryBackendNetworkComponent().getNetworkVlan().getId();
+
+            // Save the VLAN ids
+            LOG.debug("Saving VLAN details private {} and public {} for scope: {}",
+                    new Object[] { privateVlanId, publicVlanId, scopeUid });
+            savePublicVlanId(location, scopeUid, publicVlanId);
+            savePrivateVlanId(location, scopeUid, privateVlanId);
+            saveVlanDetails(machine, scopeUid, privateVlanId, publicVlanId);
+        } finally {
+            // Release the latch
+            latch.countDown();
```

Comment:
The problem with persisting latches like this is that nothing explicitly triggers the object to be written to the persisted state again. We are not triggering persistence, so it won't be written to the object store (best case, we'll get persisted some time later when the location that this is associated with next gets persisted).